query-planner-js/src/buildPlan.ts

```diff
@@ -1063,191 +1282,220 @@ class FetchDependencyGraph {
-    return this.newFetchGroup(subgraphName, entityType, true, 'query', mergeAt, directParent, pathInParent);
+    const newGroup = this.newKeyFetchGroup({ subgraphName, mergeAt });
+    newGroup.addParent(parent);
+    return newGroup
   }
 
-  newRootTypeFetchGroup(
+  removePathInParent(parent: FetchGroup, child: FetchGroup) {
+    // Simplest option is to remove the parent edge and re-add it without a path.
+    this.removeChild(parent, child);
+    child.addParent({ group: parent });
+  }
+
+  newRootTypeFetchGroup({
+    subgraphName,
+    rootKind,
+    parentType,
+    mergeAt,
+  }: {
     subgraphName: string,
     rootKind: SchemaRootKind,
     parentType: ObjectType,
     mergeAt: ResponsePath,
-    directParent: FetchGroup,
-    pathInParent: OperationPath,
-  ): FetchGroup {
-    return this.newFetchGroup(subgraphName, parentType, false, rootKind, mergeAt, directParent, pathInParent);
+  }): FetchGroup {
+    return this.newFetchGroup({ subgraphName, parentType, isEntityFetch: false, rootKind, mergeAt });
   }
 
-  // Returns true if `toCheck` is either part of `conditions`, or is a dependency (potentially recursively)
-  // of one of the gorup of conditions.
-  private isDependedOn(toCheck: FetchGroup, conditions: FetchGroup[]): boolean  {
+  // Returns true if `toCheck` is either part of `conditions`, or is one of their ancestors (potentially recursively).
+  private isInGroupsOrTheirAncestors(toCheck: FetchGroup, conditions: FetchGroup[]): boolean  {
     const stack = conditions.concat();
     while (stack.length > 0) {
       const group = stack.pop()!;
       if (toCheck.index === group.index) {
         return true;
       }
-      stack.push(...this.dependencies(group));
+      stack.push(...group.parentGroups());
     }
     return false;
   }
 
-  newKeyFetchGroup(
+  newKeyFetchGroup({
+    subgraphName,
+    mergeAt,
+  }: {
     subgraphName: string,
     mergeAt: ResponsePath,
-  ): FetchGroup {
-    const entityType = this.federationMetadata(subgraphName).entityType();
-    assert(entityType, () => `Subgraph ${subgraphName} has not entities defined`);
-    return this.newFetchGroup(subgraphName, entityType, true, 'query', mergeAt);
+  }): FetchGroup {
+    const parentType = this.federationMetadata(subgraphName).entityType();
+    assert(parentType, () => `Subgraph ${subgraphName} has not entities defined`);
+    return this.newFetchGroup({ subgraphName, parentType, isEntityFetch: true, rootKind: 'query', mergeAt });
   }
 
-  addDependency(dependentGroup: FetchGroup, dependentOn: FetchGroup | FetchGroup[], pathInDependentOn?: OperationPath) {
-    this.onModification();
-    const groups = Array.isArray(dependentOn) ? dependentOn : [ dependentOn ];
-    for (const group of groups) {
-      this.addEdge(group.index, dependentGroup.index, pathInDependentOn);
-    }
+  childs(group: FetchGroup): readonly FetchGroup[] {
+    return this.edges[group.index].outs;
```

Comment:
Is it really necessary that the FetchGroup has to know it's own index? This might be necessary for other reasons, but if it's just to be able to match the groups with the edges there is probably a better way.

Reply:
> there is probably a better way.

Isn't there always :).

More seriously, with the refactor of the previous comment, that index is mostly not used. Still, it's convenient to have it for a few reasons:
1. it's nice for debugging (`dumpOnConsole`) to associate a simple short number to each group, and index is perfect for that.
2. it's convenient when cloning a graph which we need to do because it gives us a trivial way to associate a group to its clone, which we kind of need.

Now, we can certainly achieve those things in a different way:  we could assign to each group a simple numerical id uncorrelated to the index for the first point for instance, and create some temporary map for the 2nd point. But using the `index` directly just feel easier/simpler to me at that point honestly, or at least, not worst (the only small "difficulty" of the index is that we need to adjust some of it on group removal, but it's literally 2 lines of code).